### PR TITLE
WIP/POC: handle sidecar containers

### PR DESCRIPTION
**Pull request by dicarlo2**
# Changes

Please see https://github.com/tektoncd/pipeline/issues/727 for the proposal for this POC.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
release-note
```

### CODE REVIEW

**Comment by abayer** [comment]
/ok-to-test

woo! I've been looking forward to something along these lines.

**Comment by dicarlo2** [comment]
Oh, yea, forgot to mention, not only are there no tests but I haven't even looked at the ones that already exist (and are likely failing). If we're happy with this approach I can clean up the PR + fix/add tests.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 1]
```diff
@@ -517,6 +565,23 @@ func createRedirectedTaskSpec(kubeclient kubernetes.Interface, ts *v1alpha1.Task
 			EmptyDir: &corev1.EmptyDirVolumeSource{},
 		},
 	})
+
+	// Add the downward api volume
+	ts.Volumes = append(ts.Volumes, corev1.Volume{
+		Name: entrypoint.DownwardMountName,
+		VolumeSource: corev1.VolumeSource{
+			DownwardAPI: &corev1.DownwardAPIVolumeSource{
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: i have a naiive question here about `DownwardAPIVolumeSource`, i tried to find it in the kube docs but failed: when this volume is mounted, is the content of the volume kept up to date, or is it snapshotted at mount time only? (and if it is kept up to date, how is that done? possibly relevant to our Pull Request resource plans... :innocent: )

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 2]
```diff
@@ -341,40 +359,70 @@ func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod) {
...
+				Type:    duckv1alpha1.ConditionSucceeded,
+				Status:  corev1.ConditionUnknown,
+				Reason:  "Pending",
+				Message: msg,
+			})
+		}
 	}
+
+	sidecars := len(pod.Status.ContainerStatuses) - len(taskRun.Status.Steps)
+	if pod.Status.Phase == corev1.PodRunning && sidecarsReady == sidecars {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: do you think we should worry at all about a state where a sidecar container's status indicates it is "ready", but the underlying binary isn't actually ready? (e.g. the binary is _running_ but not actually ready to handle requests)

**Comment by bobcatfish** [comment]
After our back and forth on #727 I think this approach is pretty slick, if you want to continue to flesh it out @dicarlo2 !

For next steps I think this PR in particular would need docs, tests, etc. (which I think you already know, totally understand that you wanted feedback first). Would be great to discuss at our next working group meeting on Tuesday too if you are able to join.

Anyway this seems great, let's do it! :sunglasses:

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: It's kept up to date, and afaik the kubelet handles keeping it up to date for each pod on the host by just watching the pod spec and updating appropriately.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 2 above)
Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: This approach effectively defines an API/contract for users to follow to signal to tekton that their sidecars are ready, that is, it's defined by the readiness probe of the sidecar. I think that if a use-case falls outside of the ability to be defined by a readiness probe, which can be arbitrary commands/scripts, then it's probably too complex for there to be a way to specify it inside of tekton.

Within user's steps they're free to implement any additional logic they need for more complex ready checks, this really only impacts sidecars which affect automatically specified steps (e.g. git-init). 

I think we wait until there's a clear use-case that readiness probes don't satisfy before coming up with additional ways for user's to specify readiness.

**Comment by dicarlo2** [comment]
> Would be great to discuss at our next working group meeting on Tuesday too if you are able to join.

I've put it on my schedule to attend the next one (this coming Tuesday).

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: Makes sense - sounds great :D

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: ah interesting, good to know!

**Comment by bobcatfish** [comment]
> I've put it on my schedule to attend the next one (this coming Tuesday).

I think I let you down here, I didn't add this to the agenda 🤦‍♀ (btw feel free to edit the agenda directly yourself also!) but I have NOW for the next meeting, which is Tuesday May 14 (but I won't be there unfortunately 😩 ) - the following week will be cancelled for Kubecon (any chance you'll be at kubecon?)

But long story short @dicarlo2 I think we should go ahead with this! And probably @sbwsg will need/want this for the work he's doing on log streaming :D

@dicarlo2 would you be open to having one of us continue the PR, e.g. adding some more tests, docs, the stuff we need to get this ready to merge? of course if you want to take that on that's totally cool too :D (and apologies this back and forth has taken so long!)

image

**Comment by dicarlo2** [comment]
@bobcatfish feel free to take over, I've been jumping back and forth between a few projects so I haven't had as much time as I would have liked to focus on this.

**Comment by dicarlo2** [comment]
> I think I let you down here, I didn't add this to the agenda 🤦‍♀ (btw feel free to edit the agenda directly yourself also!) but I have NOW for the next meeting, which is Tuesday May 14 (but I won't be there unfortunately 😩 ) - the following week will be cancelled for Kubecon (any chance you'll be at kubecon?)

No, you didn't let me down, as usual my schedule is a mess so I missed the meeting :/

**Comment by sbwsg** [comment]
I'm closing this PR now that @joseblas and I are working in a shared branch based off of this one.  The work can be tracked in the WIP PR here: https://github.com/tektoncd/pipeline/pull/936. Jose has done a great job getting the existing tests passing again in that branch.

**Pull request closed without merging**